service/history/mutableStateBuilder.go

```diff
@@ -1611,17 +1612,80 @@ func (e *mutableStateBuilder) CreateTransientDecisionEvents(di *decisionInfo, id
 	return scheduledEvent, startedEvent
 }
 
-func (e *mutableStateBuilder) BeforeAddDecisionTaskCompletedEvent() {
+func (e *mutableStateBuilder) beforeAddDecisionTaskCompletedEvent() {
 	// Make sure to delete decision before adding events.  Otherwise they are buffered rather than getting appended
 	e.DeleteDecision()
 }
 
-func (e *mutableStateBuilder) AfterAddDecisionTaskCompletedEvent(startedID int64) {
-	e.executionInfo.LastProcessedEvent = startedID
+func (e *mutableStateBuilder) afterAddDecisionTaskCompletedEvent(event *workflow.HistoryEvent, maxResetPoints int) {
+	e.executionInfo.LastProcessedEvent = event.GetDecisionTaskCompletedEventAttributes().GetStartedEventId()
+	e.addBinaryCheckSumIfNotExists(event, maxResetPoints)
+}
+
+// add BinaryCheckSum for the first decisionTaskCompletedID for auto-reset
+func (e *mutableStateBuilder) addBinaryCheckSumIfNotExists(event *workflow.HistoryEvent, maxResetPoints int) {
+	binChecksum := event.GetDecisionTaskCompletedEventAttributes().GetBinaryChecksum()
+	if len(binChecksum) == 0 {
+		return
+	}
+	exeInfo := e.executionInfo
+	var resetPoints map[string]*workflow.ResetPointInfo
+	if exeInfo.FeasibleAutoResetPoints != nil && exeInfo.FeasibleAutoResetPoints.ResetPoints != nil {
+		resetPoints = e.executionInfo.FeasibleAutoResetPoints.ResetPoints
+	} else {
+		resetPoints = make(map[string]*workflow.ResetPointInfo, 1)
+		e.executionInfo.FeasibleAutoResetPoints = &workflow.ResetPoints{
+			ResetPoints: resetPoints,
+		}
+	}
+
+	_, ok := resetPoints[binChecksum]
+	if ok {
+		return
+	}
+
+	if len(resetPoints) == maxResetPoints {
+		// if exceeding the max limit, do rotation by taking the oldest one out
+		var minP *workflow.ResetPointInfo
+		minK := ""
+		for k, p := range resetPoints {
```

Comment:
Can we just have an ordered data structure?  we can also keep a map for lookup but just having lookup makes this logic unnecessarily complex.

Reply:
Sure